### PR TITLE
feat: cordon pytests

### PR DESCRIPTION
**Pull request by chriswldenyer**
Python-based tests for cordon.
Some remain to be implemented because significant extra work will be needed to support their scenarios.

### CODE REVIEW

File: tests/bdd/features/cordon/node/test_cordon_node.py [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+    pytest.command_failed = False
+    ApiClient.pools_api().put_node_pool(
+        NODE_NAME_1, POOL_UUID_1, CreatePoolBody([f"{disks[0]}"])
+    )
+    yield
+    Docker.restart_container("core")
+    Cluster.cleanup()
+    remove_all_cordons(NODE_NAME_1)
+    remove_all_cordons(NODE_NAME_2)
+    delete_all_volumes()
```
Comment (tiagolobocastro) [inline tests/bdd/features/cordon/node/test_cordon_node.py]: I think cleanup deletes the volumes?
And maybe we can also change cleanup to remove cordons?

File: tests/bdd/features/cordon/node/test_cordon_node.py [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+        if os.path.exists(disk):
+            os.remove(disk)
+        with open(disk, "w") as file:
+            file.truncate(100 * 1024 * 1024)
+    # /tmp is mapped into /host/tmp within the io-engine containers
+    yield list(map(lambda file: f"/host{file}", tmp_files))
+
+    for disk in tmp_files:
+        if os.path.exists(disk):
+            os.remove(disk)
```
Comment (tiagolobocastro) [inline tests/bdd/features/cordon/node/test_cordon_node.py]: We have a new pattern now which you might want to use:
```python

@pytest.fixture(scope="module")
def disks():
    yield Deployer.create_disks(1, size=300 * 1024 * 1024)
    Deployer.cleanup_disks(1)
```

Probably need to raise a PR to refactor all tests

File: tests/bdd/features/cordon/node/test_cordon_node.py [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+
+def create_volume_and_fail(vol_uuid):
+    request = CreateVolumeBody(
+        VolumePolicy(False), NUM_VOLUME_REPLICAS, VOLUME_SIZE, False
+    )
+    try:
+        ApiClient.volumes_api().put_volume(vol_uuid, request)
+        assert False
+    except Exception as e:
+        print("failed, as expected")
```
Comment (tiagolobocastro) [inline tests/bdd/features/cordon/node/test_cordon_node.py]: nit: I'd suggest either adding more details to these prints or replacing them with `pass`

File: tests/bdd/features/cordon/node/test_cordon_node.py [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+@scenario("feature.feature", "Cordoning a node")
+def test_cordoning_a_node():
+    """Cordoning a node."""
+
+
+# Disabled because the BDD relates to kubectl idempotent behaviour which
+# differs from REST-layer behaviour - which will fail.
+# @scenario('feature.feature', 'Cordoning a node with an existing label')
+# def test_cordoning_a_node_with_an_existing_label():
+#    """Cordoning a node with an existing label."""
```
Comment (tiagolobocastro) [inline tests/bdd/features/cordon/node/test_cordon_node.py]: You can uncomment and add skip flag so it won't run

File: tests/bdd/features/cordon/node/test_cordon_node.py [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+
+def remove_all_cordons(node_name):
+    if is_cordoned(node_name):
+        node = ApiClient.nodes_api().get_node(node_name)
+        for label in node.spec.cordondrainstate["cordonedstate"]["cordonlabels"]:
+            ApiClient.nodes_api().delete_node_cordon(node_name, label)
+
+
+def share_volume(vol_uuid):
+    ApiClient.volumes_api().put_volume_share(vol_uuid, "nvmf")
```
Comment (tiagolobocastro) [inline tests/bdd/features/cordon/node/test_cordon_node.py]: Did you mean share rather than publish?
Also this doesn't seem used?

**Review by tiagolobocastro — APPROVED** [review]

File: tests/bdd/features/cordon/node/test_cordon_node.py (hunk 5 above)
Comment (chriswldenyer) [reply tests/bdd/features/cordon/node/test_cordon_node.py]: This was created and got left behind after I was trying to implement the remaining tests. I'll remove it.

File: tests/bdd/features/cordon/node/test_cordon_node.py (hunk 4 above)
Comment (chriswldenyer) [reply tests/bdd/features/cordon/node/test_cordon_node.py]: It looks like you can't skip a scenario, only "given" "when" etc.

File: tests/bdd/features/cordon/node/test_cordon_node.py (hunk 3 above)
Comment (chriswldenyer) [reply tests/bdd/features/cordon/node/test_cordon_node.py]: Added more details to the message.

File: tests/bdd/features/cordon/node/test_cordon_node.py (hunk 4 above)
Comment (chriswldenyer) [reply tests/bdd/features/cordon/node/test_cordon_node.py]: Added skip directives to disable the tests.

File: tests/bdd/features/cordon/node/test_cordon_node.py (hunk 2 above)
Comment (chriswldenyer) [reply tests/bdd/features/cordon/node/test_cordon_node.py]: fixed

File: tests/bdd/features/cordon/node/test_cordon_node.py (hunk 1 above)
Comment (chriswldenyer) [reply tests/bdd/features/cordon/node/test_cordon_node.py]: Removed the volume deletion. I'll try and get the global cordon removal added as part of the drain pytests.

File: tests/bdd/features/cordon/node/test_cordon_node.py (hunk 4 above)
Comment (tiagolobocastro) [reply tests/bdd/features/cordon/node/test_cordon_node.py]: hmm I didn't know you could skip the verbs, you can skip a scenario:
https://github.com/openebs/mayastor-control-plane/blob/4b1086adb3c05b394af1840f44a33412173ac4f4/tests/bdd/features/csi/controller/test_controller.py#L163

Comment (chriswldenyer) [reply tests/bdd/features/cordon/node/test_cordon_node.py]: Oh right. I wasn't using the right syntax. Fixed now, thanks.

**Review by Abhinandan-Purkait — APPROVED** [review]

**Comment by chriswldenyer** [comment]
bors try

**Comment by chriswldenyer** [comment]
bors try

**Comment by chriswldenyer** [comment]
bors merge

**Comment by chriswldenyer** [comment]
bors merge

**Comment by chriswldenyer** [comment]
bors merge

**Comment by tiagolobocastro** [comment]
@chriswldenyer since we enabled DCO bors merge no longer works properly when base branch changes, I think if you rebase and try bors merge again it'll work

**Comment by chriswldenyer** [comment]
bors merge

**Pull request merged**